### ISSUE TITLE
Regression after merging #73

### ISSUE THREAD

**Issue by alshakero** [Palindrom/palindrom-client#76]
This is related to Palma project.

There is a huge patch coming, that it is followed by smaller patches that are causing issues.

You'd think there is some sort of a race condition and the smaller patches assume a certain state the doesn't exist at this point and that's why the cause an exception when they're attempted to be applied. Which is not the case.

1. When I delay the applying of the smaller following patches even for 30 seconds the exception is still thrown.
2. When I delay the applying of the bigger patch itself, the smaller patches don't come! It's like the appliance of the bigger patch triggers the request for the smaller following ones, which should diminish the race condition. But once I apply the big one, the following small ones come and an exception occurs. 

I made sure that their app doesn't have the root cause by removing the panic mode (explained in #73), and it worked.

After a suggestion from @warpech I checked if the patches are same with and without the panic mode enabled, and there is a difference:
![image](https://user-images.githubusercontent.com/17054134/42501504-2762dcb0-8434-11e8-83e0-640ce89e5e9e.png)

#### Reproduction steps
1. Go to URL†.
2. Change this drop down value to "Art". 
![image](https://user-images.githubusercontent.com/17054134/42501586-50c7ac98-8434-11e8-93ab-1c41d7145ae6.png)

#### Expected outcome
Nothing, no UI changes are expected, but no console error too. 

#### Actual outcome
Nothing, no UI changes are happening, an error is thrown in the console. 

† ask for the URL and creds on Slack.

#### Setting up your environment 
1. Download [Fiddler](https://www.telerik.com/fiddler) if you don't have it.
2. Do this
![image](https://user-images.githubusercontent.com/17054134/42514283-51b02fa2-8459-11e8-80c3-7b610e25acb8.png)
3. Then this: 
![image](https://user-images.githubusercontent.com/17054134/42514385-9c1b6af2-8459-11e8-85b5-2abaf35e9e63.png)

Now you can modify the local `palindrom-client` file and see the changes.

**Comment by warpech** [comment]
I've spent some time investigating it and probably made the same full circle as @alshakero, boiling down the problem to something similar to this: https://github.com/Polymer/polymer/issues/4279

Which is a notification problem fixed in Polymer 2.

------

@alshakero perhaps it is worth trying a different approach of doing the panic mode? Instead of https://github.com/Palindrom/palindrom-client/blob/969c0f3bcf3b0cc4786aecc916d7cec985b9b691/palindrom-client.html#L234-L240, you can try something that I hacked as:

```html
var newElem = document.createElement(this.bound.localName, this.bound.getAttribute("is"));
newElem.innerHTML = '<starcounter-include view-model="{{model}}"></starcounter-include>';
nnewElem.id = this.bound.id;

var parent = this.bound.parentNode;
parent.removeChild(this.bound);
parent.appendChild(newElem);

this.bound = newElem;
this.bound.model = this.obj;
```

The idea here is to make a clone of the `bound` element as it looked before binding, remove the `bound` element and attach a fresh `bound` element. It worked well for me, except that a middle part of the current page was not re-rendered. But maybe that's due to another bug?

**Comment by alshakero** [comment]
I achieved the same outcome of this with a simpler solution, but I couldn't investigate (the empty middle area issue) deeper because it's tremendously hard to know which file contains which code, then add a Fiddler rule for it, and then play with it. 

I then got what seemed like a great idea, to find the longest common string among all patches, which refers to the root object that needs to be updated. I managed to get this path nicely in `O(N)` but when I do `domBind.set(foundPath, domBind.get(foundPath))` it is considered a write operation and the server rejects it. And when I do `notifyPath(foundPath)` nothing happens. The `notifyPath` solution would work under Polymer 2.0.

I think at this point I need access to the server code to at least modify client-side code and see if the missing middle area is a bug.

**Comment by warpech** [comment]
> I achieved the same outcome of this with a simpler solution, but I couldn't investigate (the empty middle area issue) deeper because it's tremendously hard to know which file contains which code, then add a Fiddler rule for it, and then play with it.

I find it super easy to locate any (unminified) code file using Chrome Network tab search tool. See here, searching for a URL of a HTML view that contains the word "example":

![search](https://user-images.githubusercontent.com/566463/42745934-a979111a-88d5-11e8-8b3b-2ff245a516a3.gif)

Is your case more complex?

**Comment by alshakero** [comment]
> Is your case more complex?

A little more. This whole part of the UI is coming via JSON. 

But I didn't know about this search feature. I'll find the template HTML and search for some keyword.

**Comment by alshakero** [comment]
> I achieved the same outcome of this with a simpler solution

<details>
<summary>Here is the code for the simpler solution I mentioned above if you want to play with it</summary>

```html
<!-- palindrom-polymer-client version: 7.1.0 | MIT License -->

<!--
`palindrom-client` element binds [Palindrom](https://github.com/Palindrom/Palindrom) with [Polymer's template binding](https://www.polymer-project.org/2.0/docs/devguide/templates.html).
That keeps your Polymer app, or just `dom-bind` template in sync with any server-side
data-model using Palindrom & [JSON Patch](https://tools.ietf.org/html/rfc6902) flow.
You get three-way data binding server - JS - HTML, kept in flawless sync.
    <palindrom-client
        obj="{{model}}"></palindrom-client>
It establishes the Palindrom connection when attached. All the changes made
in browser are sent to the server via WebSocket or HTTP, as
[JSON Patch](https://tools.ietf.org/html/rfc6902)es.
All the changes from server are also received and propagated to your HTML.
-->
<script src="../Palindrom/dist/palindrom-dom.min.js"></script>

<script>
    (function () {
        const defaultProps = {
            "useWebSocket": true,
            "debug": false,
            "localVersionPath": '/_ver#c$',
            "remoteVersionPath": '/_ver#s',
            "ot": true,
            "purity": false,
            "pingIntervalS": 60,
            "path": '/',
            "devToolsOpen": false,
            "fatalErrorReloadAfterS": 5,
            "mergedNotificationsThreshold": 100
        }
        const defaultAttribs = ['merged-notifications-threshold', 'remote-url', 'use-web-socket', 'debug', 'local-version-path', 'remote-version-path', 'ot', 'purity', 'listen-to', 'ping-interval-s', 'purity', 'fatal-error-reload-after-s', 'path'];
        class PalindromClient extends HTMLElement {
            constructor() {
                super();
                // assign default properties
                Object.assign(this, defaultProps);
            }
            static get observedAttributes() {
                return ['merged-notifications-threshold'];
            }
            reconnectNow() {
                this.palindrom.reconnectNow();
            }
            reload() {
                window.location.reload();
            }
            attributeChangedCallback(name, oldValue, value) {
                // it will only be called for merged-notifications-threshold attribute. See observedAttributes method
                this.mergedNotificationsThreshold = value;
            }
            cancelReloading() {
                clearInterval(this.reloadingInterval);
                delete this.reloadingInterval;
            }
            /**
             * assigns an instance of Palindrom according to given params, attaches it to .palindrom property
             *
             */
            connectedCallback() {
                // assign attributes to props
                for (const attrib of this.attributes) {
                    if (defaultAttribs.includes(attrib.name)) {
                        const prop = kebebCaseToCamelCase(attrib.name);
                        this[prop] = attrib.value;
                    }
                }
                const whereToBind = this.getAttribute("ref");
                let listenTo = this.listenTo;
                const pingIntervalS = this.pingIntervalS / 1;
                if (listenTo) {
                    listenTo = typeof listenTo == "string" ? document.getElementById(listenTo) : listenTo;
                }
                let palindrom = PalindromDOM.instances && PalindromDOM.instances.get(this.remoteUrl);
                if (!palindrom) {
                    /**
                     * To tell whether we connected to an existing instance,
                     * or created our own. This is useful for detachement, we don't want
                     * to destroy Palindrom instances that we don't own 
                     * */
                    this.isUsingOwnInstanceOfPalindrom = true;
                    this.palindrom = palindrom = new PalindromDOM({
                        remoteUrl: this.remoteUrl,
                        pingIntervalS,
                        useWebSocket: this.useWebSocket,
                        debug: this.debug,
                        ot: this.ot,
                        purity: this.purity,
                        localVersionPath: this.localVersionPath,
                        remoteVersionPath: this.remoteVersionPath,
                        listenTo,
                        onLocalChange: this.onLocalChange.bind(this),
                        onRemoteChange: this.onPatchApplied.bind(this),
                        onPatchReceived: this.onPatchReceived.bind(this),
                        onIncomingPatchValidationError: this.onGenericError.bind(this),
                        onOutgoingPatchValidationError: this.onGenericError.bind(this),
                        onError: this.onGenericError.bind(this),
                        onPatchSent: this.onPatchSent.bind(this),
                        onSocketStateChanged: this.onSocketStateChanged.bind(this),
                        onConnectionError: this.onConnectionError.bind(this),
                        onReconnectionCountdown: this.onReconnectionCountdown.bind(this),
                        onReconnectionEnd: this.onReconnectionEnd.bind(this),
                        onStateReset: obj => {
                            this.obj = obj;
                            if (whereToBind) {
                                this.bindTo(whereToBind);
                            }
                            this.onPatchApplied([{ op: 'replace', path: '', value: obj }], [{ removed: this.obj }])
                        }
                    });
                    if (!PalindromDOM.instances) {
                        PalindromDOM.instances = new Map();
                    }
                    PalindromDOM.instances.set(this.remoteUrl, palindrom);
                } else {
                    this.isUsingOwnInstanceOfPalindrom = false;
                    // with already established connection we cannot change listenTo parameter
                    this.palindrom = palindrom;
                    palindrom.onLocalChange = this.onLocalChange.bind(this);
                    palindrom.onRemoteChange = this.onPatchApplied.bind(this);
                    palindrom.onPatchReceived = this.onPatchReceived.bind(this);
                    palindrom.onIncomingPatchValidationError = this.onGenericError.bind(this);
                    palindrom.onOutgoingPatchValidationError = this.onGenericError.bind(this);
                    palindrom.onError = this.onGenericError.bind(this);
                    palindrom.onPatchSent = this.onPatchSent.bind(this);
                    palindrom.onSocketStateChanged = this.onSocketStateChanged.bind(this);
                    palindrom.onConnectionError = this.onConnectionError.bind(this);
                    palindrom.onReconnectionCountdown = this.onReconnectionCountdown.bind(this);
                    palindrom.onReconnectionEnd = this.onReconnectionEnd.bind(this);
                    palindrom.onStateReset = obj => {
                        this.obj = obj;
                        if (whereToBind) {
                            this.bindTo(whereToBind);
                        }
                        this.onPatchApplied([{ op: 'replace', path: '', value: obj }], [{ removed: this.obj }])
                    }
                    // if already established (re-)set state
                    if (palindrom.obj) {
                        palindrom.onStateReset(palindrom.obj);
                    }
                }
                this.network = palindrom.network;
                this.morphUrl = palindrom.morphUrl.bind(palindrom)
            }
            disconnectedCallback() {
                if (this.isUsingOwnInstanceOfPalindrom) {
                    this.palindrom.unlisten();
                    // hacky way to stop Palindrom connection
                    this.palindrom.useWebSocket = false;
                    PalindromDOM.instances.delete(this.remoteUrl);
                }
            }
            bindTo(element) {
                // use node id or node itself;
                element = typeof element == "string" ? document.getElementById(element) : element;
                element.model = this.obj;
                this.bound = element;
            }
            propagateDomBind(sequence) {
                if (this.bound) {
                    this.notifyTemplateDomBind(this.bound.model, sequence, this.bound, "model");
                }
                else {
                    this.notifyTemplateDomBind(this.obj, sequence, this, "obj");
                }
            }
            onLocalChange(patches) {
                // todo #26
                this.propagateDomBind({ patches, results: patches.map(() => { }) });
            }
            fire(name, detail) {
                this.dispatchEvent(new CustomEvent(name, { detail }));
            }
            onPatchReceived(data, url, method) {
                this.fire("patchreceived", {
                    data,
                    url,
                    method
                });
            }
            onPatchSent(data, url, method) {
                this.fire("patchsent", {
                    data,
                    url,
                    method
                });
            }
            onSocketStateChanged(state, url, data, statusCode, reason) {
                this.fire("socketstatechanged", {
                    state,
                    url,
                    data,
                    statusCode,
                    reason
                });
            }
            onConnectionError(palindromConnectionError) {
                const eventDetail = {
                    error: palindromConnectionError,
                    handled: false
                };
                this.fire("connection-error", eventDetail);
            }
            onGenericError(error) {
                const eventDetail = {
                    error
                };
                this.fire("generic-error", eventDetail);
            }
            onReconnectionCountdown(milliseconds) {
                const eventDetail = {
                    milliseconds,
                    handled: false
                };
                this.fire("reconnection-countdown", eventDetail);
            }
            onReconnectionEnd() {
                const eventDetail = {
                    handled: false
                };
                this.fire("reconnection-end", eventDetail);
            }
            onPatchApplied(patches, results) {
                this.propagateDomBind({ patches, results });
                this.fire("patch-applied", {
                    patches
                });
            }
            notifyTemplateDomBind(tree, patchesAndResults, templateDomBind, polymerPathPrefix) {
                if (patchesAndResults.patches.length > this.mergedNotificationsThreshold) {
                    console.warn(`palindrom-client has noticed a huge patch of ${patchesAndResults.patches.length} operations. For better performance palindrom-client will replace the whole state object instead of applying your updates individually`);
                    // clone to prevent flash of content
                    templateDomBind.set('model', {});
                    templateDomBind.set('model', this.obj);
                    return;
                }
                let operation;
                let polymerPath;
                for (let operationNo = 0, len = patchesAndResults.patches.length; operationNo < len; operationNo++) {
                    operation = patchesAndResults.patches[operationNo];
                    const result = patchesAndResults.results[operationNo];
                    if (operation.path == "") {
                        //in case of root path, Polymer replaces the original obj, which detaches it from Palindrom
                        const newSequence = { patches: [], results: [] };
                        for (const prop in operation.value) {
                            if (operation.value.hasOwnProperty(prop)) {
                                newSequence.patches.push({
                                    op: "add",
                                    path: `/${prop}`,
                                    value: operation.value[prop]
                                });
                            }
                        }
                        this.notifyTemplateDomBind(tree, newSequence, templateDomBind, polymerPathPrefix);
                        continue;
                    }
                    polymerPath = polymerPathPrefix + translateJSONPointerToPolymerPath(operation.path);
                    if (operation.op === 'test') {
                        // we assume that jsonpatch covered it already
                    } else if (operation.op === 'move' || operation.op === 'copy') {
                        console.warn("move, and copy operations are not supported yet.")
                    } else {
                        // for add and remove we need to check whether it's an array or not
                        const lastSeparator = operation.path.lastIndexOf('/');
                        let name = operation.path.substr(lastSeparator + 1);
                        const parentsPolymerPath = polymerPathPrefix + translateJSONPointerToPolymerPath(operation.path.substring(0, lastSeparator));
                        const parent = templateDomBind.get(parentsPolymerPath);
                        if (Array.isArray(parent) && (name === '-' || isNormalInteger(name))) {
                            switch (operation.op) {
                                case 'add':
                                    // JSONPatch push
                                    if (name === '-') {
                                        name = result.index;
                                    }
                                    templateDomBind.notifySplices(parentsPolymerPath,
                                        [{
                                            index: parseInt(name, 10),
                                            removed: [],
                                            addedCount: 1,
                                            object: parent,
                                            type: 'splice'
                                        }]
                                    );
                                    break;
                                case 'replace':
                                    templateDomBind.notifySplices(parentsPolymerPath,
                                        [{
                                            index: parseInt(name, 10),
                                            removed: (result && [result.removed]) || [],
                                            addedCount: 1,
                                            object: parent,
                                            type: 'splice'
                                        }]
                                    );
                                    break;
                                case 'remove':
                                    templateDomBind.notifySplices(parentsPolymerPath,
                                        [{
                                            index: parseInt(name, 10),
                                            removed: (result && [result.removed]) || [],
                                            addedCount: 0,
                                            object: parent,
                                            type: 'splice'
                                        }]
                                    );
                                    break;
                            }
                        } else {
                            switch (operation.op) {
                                case 'add':
                                case 'replace':
                                    if (typeof operation.value === 'object') {
                                        templateDomBind.notifyPath(polymerPath);
                                    } else {
                                        templateDomBind.notifyPath(polymerPath, operation.value, false);
                                    }
                                    break;
                                case 'remove':
                                    //console.warn('Polymer does not support unsetting properties https://github.com/Polymer/polymer/issues/2565');
                                    //console.warn('remove is translated to JSON incompatible set to `undefined`, as there is no remove method in Polymer.')
                                    templateDomBind.notifyPath(polymerPath, null, false);
                                    templateDomBind.notifyPath(polymerPath, undefined, false);
                                    break;
                            }
                        }
                    }
                }
            }
        }
        customElements.define('palindrom-client', PalindromClient);
        function translateJSONPointerToPolymerPath(pointer) {
            return pointer.replace(/\//g, '.');
        }
        function isNormalInteger(str) {
            const n = ~~Number(str);
            return String(n) === str && n >= 0;
        }
        function kebebCaseToCamelCase(str) {
            return str.replace(/-([a-z])/g, function (g) { return g[1].toUpperCase(); });
        }
    })();
</script>
```
</details>

---

I found out what's happening. There is a deep `dom-repeat` that populates a table that fills the missing part of the UI. This `dom-repeat` is not being re-rendereed with this fix, causing an empty UI. `notifySplice-ing` this `dom-repeat` manually fix this part of the issue. But there are other places in the UI that also need re-rendering and even notifying this `dom-repeat` alone is impossible.

Also, the current state of the code is better than this suggesion, because the same `dom-repeat` re-renders just fine with it but other issues emerge. So it's safe to say the suggesed solutions fails at earlier stage than the current code. 

However, even though current code passes this limit, it's not better. Many things are breaking badly after a huge patch is applied. I could track each. But I doesn't seem rational to fix several issues (assume they're fixable) to serve a workaround.

**Comment by warpech** [comment]
> However, even though current code passes this limit, it's not better. Many things are breaking badly after a huge patch is applied. I could track each. But I doesn't seem rational to fix several issues (assuming they're fixable) to serve a workaround.

> I suggest we rollback the fix and suggest migrating to Polymer 2.0 to get a better performance. All these fixes and time we're spending will be trashed anyways when they migrate to Polymer 2.0.

I agree. Please discuss with Erik.

**Comment by warpech** [comment]
Closing as won't fix. It was confirmed with Jimmy that they will look into migrating to Polymer 2 instead of working around Polymer 1. See https://starcounter.slack.com/archives/C67JPA7NG/p1531815185000187
